simple-request-response/README.md

```diff
@@ -0,0 +1,75 @@
+# A simple request/response abstraction
+
+| Lifecycle Stage | Maturity      | Status | Latest Revision |
+| --------------- | ------------- | ------ | --------------- |
+| 1A              | Working Draft | Active | r0, 2023-07-14  |
+
+Authors: [@MarcoPolo]
+
+todo: others welcome to join
+Interest Group: [@mxinden @marten-seemann]
+
+[@marten-seemann]: https://github.com/marten-seemann
+[@MarcoPolo]: https://github.com/MarcoPolo
+[@mxinden]: https://github.com/mxinden
+
+## Introduction
+
+Many application protocols map well to request/response semantics where a peer
+makes a request to another peer, and receives a response back. For example,
+requesting a file from a peer and getting the file back. In contrast to
+request/response, stream semantics allow two peers to continously send data
+back and forth with no specific ordering. Both styles have their uses. This
+document focuses on _one_ way of implementing request/response semantics and
+how to map it to stream based transports and an HTTP transport.
+
+## Goals of this document
+
+A goal of this document is to define _a_ simple way to write request/response
+style application protocols that can make use of all libp2p transports
+seamlessly (all stream transports and the HTTP transport). This is only one way
+of writing request/response style protocols, and users may choose to not use it.
+
+Another goal is to be backwards compatible with existing application protocols
+that follow these semantics to make upgrading them easier so that they may take
+advantage of the new HTTP transport without sacrificing backwards compatibility.
+
+The final goal is to define something simple such that it's easy to follow,
+implement, and use.
+
+## The abstraction
+
+This abstraction takes in a blob from the requesting peer (aka client) that
+represents the request and delivers the blob to the responding peer (aka
+server). The server responds with a blob that respresents the response. This
+abstraction is agnostic to what the blobs are or how they are encoded. That's up
+to the application protocol to decide.
+
+## How to run on top of a libp2p stream
+
+Each request and response should happen in a single stream. There MUST NOT be
+pipelining. After sending a request, the client SHOULD close its write side
+(signalling EOF to the peer). After handling the response, the client SHOULD
+close the stream. After sending the response the server SHOULD close the stream.
+
+## How to run on top of an HTTP transport
```

Comment:
To me, this is a separate document. Other than sharing a title, nothing that is / has been defined in this document is related to how we do this on HTTP.

Reply:
We need to define how to use this abstraction on all libp2p transports. That will soon include the HTTP transport, and thus we define that here. We could put it in a separate document "How to run simple request-response on top of an HTTP transport", but that doesn't give us anything besides adding a layer of indirection.

If we agree that "How to run on top of a libp2p stream" makes sense in this document, then "How to run on top of an HTTP transport" must also make sense in this document for the same reasons.

Reply:
> If we agree that "How to run on top of a libp2p stream" makes sense in this document, then "How to run on top of an HTTP transport" must also make sense in this document for the same reasons.

What exactly are we defining outside of these two sections?

Unless I am missing something, you are just describing what request-response is: A message in one direction with a message back, that is a pretty universally understood definition, no? Other things like disallowing of pipelining or closing of the write-side is already stream-specific.

Perhaps I am nitpicking here but I just don't see what we are actually defining that one can refer to unless you are talking about request-response on a particular transport (stream vs HTTP). Those two sections again have nothing in common so I don't quite see why they would be the same document.

Differently put, if I ask "What is request-response in libp2p if we are not talking about specific transports?", what would be your answer?

Reply:
> What exactly are we defining outside of these two sections?

What's defined in "The abstraction" section. 

>  you are just describing what request-response is

I'm not. I'm describing a very limited request-response abstraction. Notably this doesn't include metadata about the request (headers) because it aims to be backwards compatible. For example, consider some alternative request-response abstractions:

- HTTP request/response semantics
- JSON RPC
- gRPC
- [Scuttlebutt](https://ssbc.github.io/scuttlebutt-protocol-guide/#rpc-protocol)
- Simple request-response, but we add metadata via a protobuf at the beginning of the request.

>  I just don't see what we are actually defining that one can refer to 

Consider this a "thin waist layer". On one side we have user protocols that are building on top of this request response abstraction. On the other side we have the transports. Each side agrees on this abstraction and how to use it, and thus can develop independently. Application protocols building on this abstraction don't have to think about specific transports, and transports don't have to think about specific application protocols.  

> Differently put, if I ask "What is request-response in libp2p if we are not talking about specific transports?", what would be your answer?

Hopefully the above paragraph explains this a bit better. This, like the stream abstraction, is an abstraction that lets application protocols not think about specific transports.

"What is request-response in libp2p?" It is this document, or HTTP semantics, or even some user specific request-response abstraction. It's any request-response abstraction that acts as a "thin waist" between application protocols and transports.

Reply:
> > What exactly are we defining outside of these two sections?
> 
> What's defined in "The abstraction" section.

I see, so basically along the same lines as the [`Service`](https://crates.io/crates/tower-service) trait in the Rust ecosystem?